### PR TITLE
[typing/static] Execution API types

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

Add type annotations for mostly execution-related modules. Pure typing PR-- no runtime changes.

### How I Tested These Changes

BK

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12189** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12189" target="_blank">Graphite</a> 
    * **PR #12330** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12330" target="_blank">Graphite</a>  👈
      * **PR #12331** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12331" target="_blank">Graphite</a> 
        * **PR #12332** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12332" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12330?utm_source=stack-comment).

**Review by dpeng817 — COMMENTED** [review]
Test failure looks real - might need an accompanying PR in internal?

Curious about the FrozenSet -> Optional[AbstractSet...] changes, are there places where we're returning `None` for these previously / can the attributes be None?

File: examples/docs_snippets/docs_snippets/concepts/ops_jobs_graphs/op_hooks_context.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 @failure_hook
 def my_failure_hook(context: HookContext):
-    op_exception: BaseException = context.op_exception
+    op_exception: BaseException = context.op_exception  # type: ignore  # (possible none)
```
Comment (alangenfeld) [inline examples/docs_snippets/docs_snippets/concepts/ops_jobs_graphs/op_hooks_context.py:10]: I dont think having these in our docs is good - let figure out a different way

**Review by alangenfeld — CHANGES_REQUESTED** [review]

File: examples/docs_snippets/docs_snippets/concepts/ops_jobs_graphs/op_hooks_context.py (hunk 1 above)
Comment (smackesey) [reply examples/docs_snippets/docs_snippets/concepts/ops_jobs_graphs/op_hooks_context.py:10]: They are stripped out of the docs, I added that functionality [here](https://github.com/dagster-io/dagster/pull/11335)

**Comment by smackesey** [comment]
>Test failure looks real - might need an accompanying PR in internal?

Oops that already exists I just forgot to point to it in latest commit.

>Curious about the FrozenSet -> Optional[AbstractSet...] changes, are there places where we're returning None for these previously / can the attributes be None?

Yes, the existing type annotation was wrong (at runtime they can be `None`).

File: examples/docs_snippets/docs_snippets/concepts/ops_jobs_graphs/op_hooks_context.py (hunk 1 above)
Comment (alangenfeld) [reply examples/docs_snippets/docs_snippets/concepts/ops_jobs_graphs/op_hooks_context.py:10]: ah cool

File: python_modules/dagster/dagster/_core/executor/in_process.py [hunk 2]
```diff
@@ -47,7 +50,7 @@ def execute(self, plan_context, execution_plan):
                         pipeline_run=plan_context.pipeline_run,
                         instance=plan_context.instance,
                         raise_on_error=plan_context.raise_on_error,
-                        output_capture=plan_context.output_capture,
+                        output_capture=plan_context.output_capture,  # type: ignore  # (??)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/executor/in_process.py:53]: what was this about

File: python_modules/dagster/dagster/_core/storage/pipeline_run.py [hunk 3]
```diff
@@ -645,23 +646,25 @@ def __new__(
         )
 
     @property
-    def pipeline_name(self):
-        return self.job_name
+    def pipeline_name(self) -> str:
+        return self.job_name  # type: ignore  # (possible none)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/pipeline_run.py:650]: this should be typed as optional - filtering on job name is optional

File: python_modules/dagster/dagster/_core/execution/resolve_versions.py [hunk 4]
```diff
@@ -110,10 +110,10 @@ def resolve_step_versions(
         solid_config = resolved_run_config.solids[solid_name].config
 
         solid_def_version = None
-        if solid_def.version is not None:
-            solid_def_version = solid_def.version
+        if solid_def.version is not None:  # type: ignore  # (should be OpDefinition)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/execution/resolve_versions.py:113]: what is `solid_def` instead? do we just need to update the call above to a different getter?

**Review by alangenfeld — APPROVED** [review]
Im pretty sure the `RunsFilter` change should be done - others worth taking a second peek

File: python_modules/dagster/dagster/_core/execution/resolve_versions.py (hunk 4 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/execution/resolve_versions.py:113]: It's `NodeDefinition`, which doesn't have `version`-- will fix this upstack

Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/execution/resolve_versions.py:113]: i dont think we should get intermediate Graph nodes here so probably can use whatever getter ensures its returning an `OpDefinition`

File: python_modules/dagster/dagster/_core/executor/in_process.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/executor/in_process.py:53]: fixed

File: python_modules/dagster/dagster/_core/storage/pipeline_run.py (hunk 3 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/storage/pipeline_run.py:650]: good catch

**Pull request merged**